library/pk_wrap.c

```diff
@@ -716,4 +720,214 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+    mbedtls_free( ctx );
+}
+
+static size_t pk_psa_get_bitlen( const void *ctx )
+{
+    const psa_key_slot_t *key = (const psa_key_slot_t *) ctx;
+    size_t bits;
+
+    if( PSA_SUCCESS != psa_get_key_information( *key, NULL, &bits ) )
+        return( 0 );
+
+    return( bits );
+}
+
+static int pk_psa_can_do( mbedtls_pk_type_t type )
+{
+    /* For now opaque PSA keys can only wrap ECC keypairs,
+     * as checked by setup_psa().
+     * Also, ECKEY_DH does not really make sense with the current API. */
+    return( type == MBEDTLS_PK_ECKEY ||
+            type == MBEDTLS_PK_ECDSA );
+}
+
+/* Note: duplicated from Andrzej's PR as a quick fix. Needs to be deduplicated
+ * before merging */
+static psa_algorithm_t translate_md_to_psa( mbedtls_md_type_t md_alg )
+{
+    switch( md_alg )
+    {
+#if defined(MBEDTLS_MD2_C)
+    case MBEDTLS_MD_MD2:
+        return( PSA_ALG_MD2 );
+#endif
+#if defined(MBEDTLS_MD4_C)
+    case MBEDTLS_MD_MD4:
+        return( PSA_ALG_MD4 );
+#endif
+#if defined(MBEDTLS_MD5_C)
+    case MBEDTLS_MD_MD5:
+        return( PSA_ALG_MD5 );
+#endif
+#if defined(MBEDTLS_SHA1_C)
+    case MBEDTLS_MD_SHA1:
+        return( PSA_ALG_SHA_1 );
+#endif
+#if defined(MBEDTLS_SHA256_C)
+    case MBEDTLS_MD_SHA224:
+        return( PSA_ALG_SHA_224 );
+    case MBEDTLS_MD_SHA256:
+        return( PSA_ALG_SHA_256 );
+#endif
+#if defined(MBEDTLS_SHA512_C)
+    case MBEDTLS_MD_SHA384:
+        return( PSA_ALG_SHA_384 );
+    case MBEDTLS_MD_SHA512:
+        return( PSA_ALG_SHA_512 );
+#endif
+#if defined(MBEDTLS_RIPEMD160_C)
+    case MBEDTLS_MD_RIPEMD160:
+        return( PSA_ALG_RIPEMD160 );
+#endif
+    case MBEDTLS_MD_NONE:  // Intentional fallthrough
+    default:
+        return( 0 );
+    }
+}
+
+/* Like mbedtls_asn1_write_mpi, but from a buffer */
+static int asn1_write_mpibuf( unsigned char **p, unsigned char *start,
+                              const unsigned char *src, size_t slen )
+{
+    int ret;
+    size_t len = 0;
+
+    if( (size_t)( *p - start ) < slen )
+        return( MBEDTLS_ERR_ASN1_BUF_TOO_SMALL );
+
+    len = slen;
+    *p -= len;
+    memcpy( *p, src, len );
+
+    if( **p & 0x80 )
+    {
+        if( *p - start < 1 )
+            return( MBEDTLS_ERR_ASN1_BUF_TOO_SMALL );
+
+        *--(*p) = 0x00;
+        len += 1;
+    }
+
+    MBEDTLS_ASN1_CHK_ADD( len, mbedtls_asn1_write_len( p, start, len ) );
+    MBEDTLS_ASN1_CHK_ADD( len, mbedtls_asn1_write_tag( p, start, MBEDTLS_ASN1_INTEGER ) );
+
+    return( (int) len );
+}
+
+/* Transcode signature from PSA format to ASN.1 sequence.
+ * See ecdsa_signature_to_asn1 in ecdsa.c.
+ *
+ * [in] sig: the signature in PSA format
+ * [in/out] sig_len: signature length pre- and post-transcoding
+ * [out] dst: the signature in ASN.1 format
+ */
+static int pk_ecdsa_sig_asn1_from_psa( const unsigned char *sig, size_t *sig_len,
+                                       unsigned char *dst )
+{
+    int ret;
+    unsigned char buf[MBEDTLS_ECDSA_MAX_LEN];
```

Comment:
I think we should avoid such large stack allocations and work in place on the target buffer.

Reply:
In principle I would agree, but unfortunately we don't know the size of the output buffer, so it we want to use it directly we can't write to its end and then to a `memmove()`, we'd have to first compute the length of the signature, then write it starting from its end. It would make the code more complex and harder to maintain (need to keep in sync the length computation with the ASN.1 writing code).

Also, please note that so far we already had such a buffer in `ecdsa_signature_to_asn1()`, which is not used by `psa_ecdsa_sign()`, so compared to the non-PSA way, the increase of stack usage (disregarding frames and small local variables) is from `MBEDTLS_ECDSA_MAX_LEN` to `MBEDTLS_ECDSA_MAX_LEN + 2 * MBEDTLS_ECP_MAX_BYTES`. So an alternative to what I wrote above would allocate a single buffer of size `MBEDTLS_ECDSA_MAX_LEN` in `pk_psa_sign_wrap()` and then make `pk_ecdsa_sig_asn1_from_psa` and still `memcpy` to the actual destination buffer at the end. This would avoid the complexity of computing length before writing, and keep usage of large stack buffers in line with the non-PSA case.

Please let me know what you think.

Reply:
> In principle I would agree, but unfortunately we don't know the size of the output buffer, so it we want to use it directly we can't write to its end and then to a memmove(), we'd have to first compute the length of the signature, then write it starting from its end. It would make the code more complex and harder to maintain (need to keep in sync the length computation with the ASN.1 writing code).

I agree it's a bit tedious, but I think it's worth it. Also, I think one could even get rid of _any_ stack buffers by calling `psa_ecdsa_sign()` on the destination buffer (which we assume is large enough to hold the ASN.1 version of the signature, and so it must also be large enough for the raw one), and then modify `ecdsa_signature_to_asn1()` to work in place.

Reply:
I got rid of both stack buffers. My solution is based on pre-computing the maximal length of any signature for this curve size, rather than the actual size of this particular signature, so there's a `memmove()` call near the end of `pk_ecdsa_sig_asn1_from_psa()` that could be avoided, but I think it's an acceptable compromise.